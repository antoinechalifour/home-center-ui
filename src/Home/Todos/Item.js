import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { graphql, compose } from 'react-apollo'
import gql from 'graphql-tag'
import TodoIcon from 'react-icons/lib/md/check-box-outline-blank'
import DoneIcon from 'react-icons/lib/md/check-box'
import DeleteIcon from 'react-icons/lib/md/clear'

const Wrapper = styled.li`
  display: flex;
  align-items: center;
  padding: 12px;
  cursor: pointer;

  svg {
    margin-right: 12px;
    font-size: 20px;
  }

  span {
    flex: 1;
    ${({ checked }) => checked && 'text-decoration: line-through;'}
    ${({ checked }) => checked && 'opacity: .54;'}
  }

  span + svg {
    margin-right: 0;
    opacity: .54;
  }

  + li {
    border-top: 1px solid rgba(0, 0, 0, .15);
  }
`

const Item = ({ id, text, done, updateListItem, deleteListItem }) => (
  <Wrapper
    checked={done}
    onClick={() => updateListItem({ variables: { id, text, done: !done } })}
  >
    {done ? <DoneIcon /> : <TodoIcon />}
    <span>{text}</span>
    <DeleteIcon onClick={() => deleteListItem({ variables: { id } })} />
  </Wrapper>
)

Item.propTypes = {
  id: PropTypes.number.isRequired,
  text: PropTypes.string.isRequired,
  done: PropTypes.bool.isRequired,
  updateListItem: PropTypes.func.isRequired,
  deleteListItem: PropTypes.func.isRequired
}

export default compose(
  graphql(
    gql`
      mutation updateListItem ($id: Int, $text: String, $done: Boolean) {
        updateListItem(id: $id, text: $text, done: $done) {
          id
        }
      }
    `,
    {
      name: 'updateListItem',
      options: {
        refetchQueries: ['ListQuery']
      }
    }
  ),
  graphql(
    gql`
      mutation deleteListItem ($id: Int) {
        deleteListItem(id: $id) {
          id
        }
      }
    `,
    {
      name: 'deleteListItem',
      options: {
        refetchQueries: ['ListQuery']
      }
    }
  )
)(Item)
