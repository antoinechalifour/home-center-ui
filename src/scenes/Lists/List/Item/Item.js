import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import TodoIcon from 'react-icons/lib/md/check-box-outline-blank'
import DoneIcon from 'react-icons/lib/md/check-box'
import DeleteIcon from 'react-icons/lib/md/clear'
import Editable from 'components/Editable'

export default function Item ({
  id,
  text,
  done,
  updateListItem,
  deleteListItem
}) {
  const toggleStatus = () =>
    updateListItem({ variables: { id, text, done: !done } })
  const CheckboxIcon = done ? DoneIcon : TodoIcon
  const updateText = value =>
    updateListItem({ variables: { id, text: value, done } })

  return (
    <Wrapper checked={done}>
      <CheckboxIcon onClick={toggleStatus} />
      <Editable onChange={updateText}>{text}</Editable>
      <DeleteIcon onClick={() => deleteListItem({ variables: { id } })} />
    </Wrapper>
  )
}

const Wrapper = styled.li`
  display: flex;
  align-items: center;
  padding: 12px;

  svg {
    font-size: 20px;
    cursor: pointer;
  }

  svg:first-child {
    margin-right: 12px;
  }

  span {
    flex: 1;
    ${({ checked }) => checked && 'text-decoration: line-through;'}
    ${({ checked }) => checked && 'opacity: .54;'}
  }

  svg:last-child {
    margin-left: 12px;
    opacity: .54;
  }

  + li {
    border-top: 1px solid rgba(0, 0, 0, .15);
  }
`

Item.propTypes = {
  id: PropTypes.number.isRequired,
  text: PropTypes.string.isRequired,
  done: PropTypes.bool.isRequired,
  updateListItem: PropTypes.func.isRequired,
  deleteListItem: PropTypes.func.isRequired
}
