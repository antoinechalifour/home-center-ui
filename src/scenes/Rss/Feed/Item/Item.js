import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import DeleteIcon from 'react-icons/lib/md/delete'
import ExpandableMenu, { MenuItem } from 'ui/ExpandableMenu'

export default function Item ({ title, link, source, sourceId, deleteSource }) {
  return (
    <Container>
      <div flex>
        <Link href={link} target='_blank'>
          <span>{title}</span>
          <ItemMeta>{source}</ItemMeta>
        </Link>
      </div>
      <Menu
        renderItems={() => (
          <div>
            <MenuItem onClick={deleteSource}>
              <DeleteIcon /><span>Delete source</span>
            </MenuItem>
          </div>
        )}
      />
    </Container>
  )
}

Item.propTypes = {
  title: PropTypes.string.isRequired,
  link: PropTypes.string.isRequired,
  source: PropTypes.string.isRequired,
  sourceId: PropTypes.number.isRequired,
  deleteSource: PropTypes.func.isRequired
}

const Container = styled.div`
  display: flex;
`

const Link = styled.a`
  box-sizing: border-box;
  padding: 12px;
  display: block;
  height: 100%;
  overflow: hidden;
  text-decoration: none;
  color: inherit;

  span {
    display: block;
  }
`

const ItemMeta = styled.span`
  font-size: 80%;
  opacity: .54;
`

const Menu = styled(ExpandableMenu)`
  svg {
    opacity: .54;
  }
`
