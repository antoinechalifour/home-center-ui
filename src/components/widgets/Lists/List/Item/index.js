import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { graphql, compose } from 'react-apollo'
import { updateListItem, deleteListItem } from 'queries/lists'
import Item from './Item'

class ItemContainer extends Component {
  static propTypes = {
    id: PropTypes.number.isRequired,
    text: PropTypes.string.isRequired,
    done: PropTypes.bool.isRequired,
    updateListItem: PropTypes.func.isRequired,
    deleteListItem: PropTypes.func.isRequired
  }

  _onToggleStatus = () => {
    this.props.updateListItem({
      variables: {
        input: {
          id: this.props.id,
          text: this.props.text,
          done: !this.props.done
        }
      }
    })
  }

  _onUpdateText = value => {
    this.props.updateListItem({
      variables: {
        input: {
          id: this.props.id,
          text: value,
          done: this.props.done
        }
      }
    })
  }

  _onDelete = () => {
    this.props.deleteListItem({
      variables: {
        input: { id: this.props.id }
      }
    })
  }

  render () {
    return (
      <Item
        {...this.props}
        updateText={this._onUpdateText}
        toggleStatus={this._onToggleStatus}
        deleteItem={this._onDelete}
      />
    )
  }
}

const updateListItemOptions = {
  name: 'updateListItem'
}

const deleteListItemOptions = {
  name: 'deleteListItem'
}

export default compose(
  graphql(updateListItem, updateListItemOptions),
  graphql(deleteListItem, deleteListItemOptions)
)(ItemContainer)
