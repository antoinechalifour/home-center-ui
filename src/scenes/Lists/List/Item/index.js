import { graphql, compose } from 'react-apollo'
import { updateListItem, deleteListItem } from 'queries/lists'
import Item from './Item'

const updateListItemOptions = {
  name: 'updateListItem',
  options: {
    refetchQueries: ['GetList']
  }
}

const deleteListItemOptions = {
  name: 'deleteListItem',
  options: {
    refetchQueries: ['GetList']
  }
}

export default compose(
  graphql(updateListItem, updateListItemOptions),
  graphql(deleteListItem, deleteListItemOptions)
)(Item)
