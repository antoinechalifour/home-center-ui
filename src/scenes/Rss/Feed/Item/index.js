import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { graphql } from 'react-apollo'
import { deleteSource } from 'queries/rss'
import Item from './Item'

class ItemContainer extends Component {
  static propTypes = {
    sourceId: PropTypes.number.isRequired,
    mutate: PropTypes.func.isRequired
  }

  deleteSource = () => {
    this.props.mutate({
      variables: {
        input: { id: this.props.sourceId }
      }
    })
  }

  render () {
    return <Item {...this.props} deleteSource={this.deleteSource} />
  }
}

const queryOptions = {
  options: {
    refetchQueries: ['GetRssQuery']
  }
}

export default graphql(deleteSource, queryOptions)(ItemContainer)
