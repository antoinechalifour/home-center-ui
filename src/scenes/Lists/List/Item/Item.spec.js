jest.mock('react-icons/lib/md/check-box-outline-blank', () => 'UncheckedIcon')
jest.mock('react-icons/lib/md/check-box', () => 'CheckedIcon')
jest.mock('react-icons/lib/md/clear', () => 'DeleteIcon')

import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import { shallow } from 'enzyme'
import Item from './Item'

const renderer = new ShallowRenderer()

describe('scenes/Lists/List/Item', () => {
  it('Should render a unchecked icon when the item is not done', () => {
    const props = {
      id: 2,
      text: 'This is an item',
      done: false,
      updateListItem: jest.fn(),
      deleteListItem: jest.fn()
    }
    const tree = renderer.render(<Item {...props} />)

    expect(tree).toMatchSnapshot()
  })

  it('Should render a checked icon when the item is done', () => {
    const props = {
      id: 2,
      text: 'This is an item',
      done: true,
      updateListItem: jest.fn(),
      deleteListItem: jest.fn()
    }
    const tree = renderer.render(<Item {...props} />)

    expect(tree).toMatchSnapshot()
  })

  it('Should call the delete mutation when the delete icon is clicked', () => {
    const props = {
      id: 2,
      text: 'This is an item',
      done: true,
      updateListItem: jest.fn(),
      deleteListItem: jest.fn()
    }
    const wrapper = shallow(<Item {...props} />)
    wrapper.find('DeleteIcon').simulate('click')

    expect(props.deleteListItem.mock.calls.length).toBe(1)
    expect(props.deleteListItem.mock.calls[0][0]).toEqual({
      variables: {
        id: 2
      }
    })
  })

  it('Should call the update mutation when the checkbox icon is clicked', () => {
    const props = {
      id: 2,
      text: 'This is an item',
      done: true,
      updateListItem: jest.fn(),
      deleteListItem: jest.fn()
    }
    const wrapper = shallow(<Item {...props} />)
    wrapper.find('CheckedIcon').simulate('click')

    expect(props.updateListItem.mock.calls.length).toBe(1)
    expect(props.updateListItem.mock.calls[0][0]).toEqual({
      variables: {
        id: 2,
        text: props.text,
        done: false
      }
    })
  })
})
