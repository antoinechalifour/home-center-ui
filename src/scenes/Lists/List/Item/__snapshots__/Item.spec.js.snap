// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scenes/Lists/List/Item Should render a checked icon when the item is done 1`] = `
<styled.li
  checked={true}
>
  <CheckedIcon
    onClick={[Function]}
  />
  <span>
    This is an item
  </span>
  <DeleteIcon
    onClick={[Function]}
  />
</styled.li>
`;

exports[`scenes/Lists/List/Item Should render a unchecked icon when the item is not done 1`] = `
<styled.li
  checked={false}
>
  <UncheckedIcon
    onClick={[Function]}
  />
  <span>
    This is an item
  </span>
  <DeleteIcon
    onClick={[Function]}
  />
</styled.li>
`;
